SourceCode/HW2/HW2/HW2.cpp

```diff
@@ -0,0 +1,174 @@
+#include <fstream>
+#include <string>
+#include <array>
+#include <vector>
+#include <random>
+#include <numbers>
+#include <iostream>
+
+/// Output image resolution
+static constexpr uint32_t imageWidth = 1000;
+static constexpr uint32_t imageHeight = 1000;
+static constexpr uint32_t maxColorComponent = 255;
+
+class PPMWriter 
+{
+public:
+
+	PPMWriter(std::string filename, uint32_t imageWidth, uint32_t imageHeight, uint32_t maxColorComponent)
+		: ppmFileStream(filename + ".ppm", std::ios::out | std::ios::binary)
+	{
+		ppmFileStream << "P3\n";
+		ppmFileStream << imageWidth << " " << imageHeight << "\n";
+		ppmFileStream << maxColorComponent << "\n";
+	}
+
+	~PPMWriter() 
+	{
+		ppmFileStream.close();
+	}
+
+	friend PPMWriter& operator<<(PPMWriter& writer, const std::string& data) 
+	{
+		writer.ppmFileStream << data;
+		return writer;
+	}
+
+private:
+	std::ofstream ppmFileStream;
+};
+
+class RGB 
+{
+public:
+	uint8_t r;
+	uint8_t g;
+	uint8_t b;
+
+	std::string ToString() const
+	{
+		return std::to_string(r) + " " + std::to_string(g) + " " + std::to_string(b);
+	}
+};
+
+class HSV
+{
+public:
+	float h; // Hue in degrees [0, 360)
+	float s; // Saturation as a percentage [0, 100]
+	float v; // Value as a percentage [0, 100]
+
+	HSV(float hue, float saturation, float value)
+		: h(hue), s(saturation), v(value) {}
+
+	RGB ToRGB() const
+	{
+		float r, g, b;
+
+		float s_norm = s / 100.0f;
+		float v_norm = v / 100.0f;
+		float c = v_norm * s_norm;
+		float x = c * (1 - fabs(fmod(h / 60.0, 2) - 1));
+		float m = v_norm - c;
+
+		if (h >= 0 && h < 60) {
+			r = c, g = x, b = 0;
+		}
+		else if (h >= 60 && h < 120) {
+			r = x, g = c, b = 0;
+		}
+		else if (h >= 120 && h < 180) {
+			r = 0, g = c, b = x;
+		}
+		else if (h >= 180 && h < 240) {
+			r = 0, g = x, b = c;
+		}
+		else if (h >= 240 && h < 300) {
+			r = x, g = 0, b = c;
+		}
+		else {
+			r = c, g = 0, b = x;
+		}
+
+		return RGB{ static_cast<uint8_t>((r + m) * 255), static_cast<uint8_t>((g + m) * 255), static_cast<uint8_t>((b + m) * 255) };
+	}
+};
+
+float rnd() 
+{
+	static std::random_device rd;  // Seed
+	static std::mt19937 gen(rd()); // Mersenne Twister generator
+	std::uniform_real_distribution<> dis(0.0, 1.0);
+	return dis(gen);
+}
+
+float frac(float number) 
+{
+	return number - std::floor(number);
+}
+
+void GenerateRectangles(uint32_t verticalSegments, uint32_t horizontalSegments)
+{
+	const uint32_t colorCount = horizontalSegments * verticalSegments;
+
+	// Generate low-discrepancy colors using golden ratio
```

Comment:
I like the idea of that, just keep in mind that generating the colors in this way could introduce patterns that you might not want. You might notice it more with different number of segments.

Reply:
Yes, I wanted to create a simple low-discrepancy sequence of colors, and I thought of using the golden ratio trick. It's a complex topic overall, and I didn't want to delve deeper. :D